### PR TITLE
Bspline dict2

### PR DESCRIPTION
**Pull request by EmAstro**
Updated from_dict() as suggested by Kyle.
Fixed: np.array -> list and vice versa

### CODE REVIEW

File: pypeit/core/pydl.py [hunk 1]
```diff
@@ -233,112 +233,137 @@ def __init__(self, x, nord=4, npoly=1, bkpt=None, fullbkpt = None, bkspread=1.0,
...
+            self.coeff = np.zeros((nc,), dtype='d')
+            self.icoeff = np.zeros((nc,), dtype='d')
+        self.xmin = 0.0
+        self.xmax = 1.0
+        if 'funcname' in kwargs:
+            self.funcname = kwargs['funcname']
+        else:
+            self.funcname = 'legendre'
+
+    @classmethod
```
Comment (jhennawi) [inline pypeit/core/pydl.py]: Hi @EmAstro, 

It does not make sense for 'x' to be an argument here. The point is that if we are instantiating from a dict, then there is no independent variable provided as it has already been provided previously to create the bspline that we are trying to restore. Note that the bspline object does not have an 'x' attribute, but rather upon instantiation it uses the independent variable x to determine its attributes. 

Go ahead and remove the x from the argument list and return list and I will approve this.

**Comment by EmAstro** [comment]
updated as requested

File: pypeit/core/pydl.py [hunk 2]
```diff
@@ -233,112 +233,136 @@ def __init__(self, x, nord=4, npoly=1, bkpt=None, fullbkpt = None, bkspread=1.0,
...
+            bspline dictionary created with to_dict().
+
+        Returns
+        -------
+        a bspline object
         """
-        Write bspline parameters to a dict.
+
+        # needs to move lists to np.arrays for JSON files
+        return cls(nord=from_dict['nord'],
```
Comment (kbwestfall) [inline pypeit/core/pydl.py]: This won't work without the `x`.  You can just give `x` as None, though, as in:

```
return cls(None, nord=d['nord'], ...)
```

File: pypeit/core/pydl.py [hunk 3]
```diff
@@ -233,112 +233,136 @@ def __init__(self, x, nord=4, npoly=1, bkpt=None, fullbkpt = None, bkspread=1.0,
...
+            self.icoeff = np.zeros((nc,), dtype='d')
+        self.xmin = 0.0
+        self.xmax = 1.0
+        if 'funcname' in kwargs:
+            self.funcname = kwargs['funcname']
+        else:
+            self.funcname = 'legendre'
+
+    @classmethod
+    def from_dict(cls, from_dict):
```
Comment (kbwestfall) [inline pypeit/core/pydl.py]: I wouldn't call the name of the function and the argument of the function the same thing.  I.e., I would change this to, e.g., `def from_dict(cls, input):`

File: pypeit/core/pydl.py [hunk 4]
```diff
@@ -233,112 +233,136 @@ def __init__(self, x, nord=4, npoly=1, bkpt=None, fullbkpt = None, bkspread=1.0,
...
+
+        # needs to move lists to np.arrays for JSON files
+        return cls(nord=from_dict['nord'],
+                   npoly=from_dict['npoly'],
+                   bkpt=np.array(from_dict['bkpt']),
+                   mask=np.array(from_dict['mask']),
+                   coeff=np.array(from_dict['coeff']),
+                   icoeff=np.array(from_dict['icoeff']),
+                   xmin=from_dict['xmin'],
+                   xmax=from_dict['xmax'],
```
Comment (kbwestfall) [inline pypeit/core/pydl.py]: Looking at the way the `__init__` function uses `x`, it needs its minimum, maximum, and length.  Given that this `from_dict` method uses `xmin` and `xmax` as keywords, I would remove `x` as an argument and add its size.  I.e.: `def __init__(self, xmin=None, xmax=None, xsize=None, ...)`

File: pypeit/core/pydl.py (hunk 2 above)
Comment (jhennawi) [reply pypeit/core/pydl.py]: Why want it work without an argument. I'm confused?

Comment (kbwestfall) [reply pypeit/core/pydl.py]: Right, the arguments sent to cls have to be the same as you would use to instantiate the object.  I.e., you have to give it all the arguments (but not keyworded arguments) of the `__init__` method, except `self`.  You could change the `__init__` to: `def __init__(x=None, nord=4, ...)`, or you can add a None to this class method, `return cls(None, nord=d['nord'], ...)`.

**Comment by profxj** [comment]
@EmAstro  -- Let's move this PR along.  
Should prefer not to leave ones like this lingering.

**Comment by jhennawi** [comment]
Hi @kbwestfall, 

The spicy bspline is pretty crappy. Xavier was using that before and it was causing all kinds of silly artifacts. The pydl bspline has a very sophisticated front end that has been tested by decades of use. That is why we have a custom one. All the extraction is built on that, and there is no going back. I am all using native python scipy/numpy things, except in cases like this where those native things stink. I have ideas on how to make the pydl bspline more robust, for example currently the matrix operations it performs are coded in pydl, but we should be doing these computations in scipy/numpy.  Anyway, it is all coded in python at the moment, although admittedly some of that code is not that elegant. 

You are right about the x vector. It should be kept in the class. I suggest you make that an issue. 

The reason we want the from_dict and to_dict methods is that we need to write these objects out to a file in some instances, and there seems to be no more robust way to do this, than convert to a dict and json write the dict. The reason we need to write them to a file is because for example this is the result of the sensitivity function computation and we need to save that as a master file. 

Joe

**Comment by kbwestfall** [comment]
Merging.

**Pull request merged**